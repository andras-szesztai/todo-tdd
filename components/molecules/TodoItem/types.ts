export interface TTodoItem {
    id: string
    createdAt: string
    title: string
    isCompleted: boolean
}

export interface Props extends Pick<TTodoItem, 'title' | 'isCompleted'> {
    onCompleteClick: () => void
}
