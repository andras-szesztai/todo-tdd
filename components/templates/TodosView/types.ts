import { TTodoItem } from 'components/molecules/TodoItem'

export interface Props {
    initialTodoItems?: TTodoItem[]
}
