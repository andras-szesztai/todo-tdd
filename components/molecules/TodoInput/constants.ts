export const TODO_INPUT_LABEL = 'Todo Title'
export const TODO_INPUT_PLACEHOLDER = 'Enter a Todo title...'
export const TODO_INPUT_BUTTON_TEXT = 'Submit'

export const TODO_INPUT_ID = 'todo-input'
