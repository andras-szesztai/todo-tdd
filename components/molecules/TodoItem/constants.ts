export const TODO_COMPLETE_BUTTON_TEXT = {
    complete: 'Complete',
    unComplete: 'Un-complete',
}
