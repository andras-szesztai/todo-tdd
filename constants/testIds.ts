export const TEST_ID = {
    todoInputLabel: 'todo-input-label',
    todoInput: 'todo-input',
    todoInputButton: 'todo-input-button',
    todoViewContainer: 'todos-view-container',
    todoListContainer: 'todo-list-container',
    todoItem: 'todo-item',
    todoCompleteButton: 'todo-complete-button',
}
