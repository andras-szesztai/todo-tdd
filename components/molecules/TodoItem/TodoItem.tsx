import { useState } from 'react'
import classNames from 'classnames'

import { Button } from 'components/atoms/Button'

import { TODO_COMPLETE_BUTTON_TEXT } from './constants'
import { Props } from './types'

function TodoItem({ title, onCompleteClick, isCompleted }: Props) {
    const liStyles = classNames(
        'relative border border-primary-dark rounded-lg p-2 truncate group',
        {
            'bg-transparent border-primary-dark-disabled text-white':
                isCompleted,
            'bg-white border-primary-dark text-primary-dark': !isCompleted,
        }
    )
    const [isFocused, setIsFocused] = useState(false)
    const spanStyles = classNames(
        'absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100',
        { 'opacity-100': isFocused }
    )
    return (
        <li className={liStyles}>
            {title}
            <span className={spanStyles}>
                <Button
                    text={
                        isCompleted
                            ? TODO_COMPLETE_BUTTON_TEXT.unComplete
                            : TODO_COMPLETE_BUTTON_TEXT.complete
                    }
                    onClick={onCompleteClick}
                    size="small"
                    onFocus={() => {
                        setIsFocused(true)
                    }}
                    onBlur={() => {
                        setIsFocused(false)
                    }}
                />
            </span>
        </li>
    )
}

export default TodoItem
