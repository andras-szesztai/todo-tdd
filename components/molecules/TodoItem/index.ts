import TodoItem from './TodoItem'
import { TTodoItem } from './types'
import { TODO_COMPLETE_BUTTON_TEXT } from './constants'

export { TodoItem, TODO_COMPLETE_BUTTON_TEXT }
export type { TTodoItem }
