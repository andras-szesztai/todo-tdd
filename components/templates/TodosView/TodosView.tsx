'use client'

import { useState } from 'react'
import { v4 as uuidv4 } from 'uuid'

import { TodoInput } from 'components/molecules/TodoInput'
import { TTodoItem } from 'components/molecules/TodoItem'
import { TodoList } from 'components/organisms/TodoList'
import { TEST_ID } from 'constants/testIds'
import { Props } from './types'

function TodosView({ initialTodoItems = [] }: Props) {
    const [todos, setTodos] = useState<TTodoItem[]>(initialTodoItems)

    const handleTodoItemCompleteClick = (id: string) => {
        const nextTodos = [...todos]
        const todoIndex = nextTodos.findIndex((todo) => todo.id === id)
        nextTodos[todoIndex].isCompleted = !nextTodos[todoIndex].isCompleted
        setTodos(nextTodos)
    }

    return (
        <section
            data-testid={TEST_ID.todoViewContainer}
            className="flex flex-col gap-4 font-medium -translate-y-12 text-primary-dark"
        >
            <TodoInput
                onSubmit={(newTodo) => {
                    setTodos((prev) => [
                        ...prev,
                        {
                            id: uuidv4(),
                            title: newTodo,
                            createdAt: new Date().toISOString(),
                            isCompleted: false,
                        },
                    ])
                }}
            />
            <TodoList
                todoItems={todos}
                onTodoItemCompleteClick={handleTodoItemCompleteClick}
            />
        </section>
    )
}

export default TodosView
