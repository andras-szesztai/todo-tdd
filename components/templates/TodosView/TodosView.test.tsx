import { fireEvent, render, within } from '@testing-library/react'
import { TODO_COMPLETE_BUTTON_TEXT } from 'components/molecules/TodoItem'
import { TEST_ID } from 'constants/testIds'

import TodosView from './TodosView'

describe('TodosView', () => {
    it('renders without error', () => {
        const { getByTestId } = render(<TodosView />)
        expect(getByTestId(TEST_ID.todoViewContainer)).toBeInTheDocument()
    })

    it('displays submitted todos in todos list, todos with same name can be submitted', () => {
        const { getByRole, queryByTestId, getAllByTestId, asFragment } = render(
            <TodosView />
        )
        expect(asFragment()).toMatchSnapshot()
        expect(queryByTestId(TEST_ID.todoItem)).not.toBeInTheDocument()
        const input = getByRole('textbox')
        const firstItem = 'My first todo'
        fireEvent.change(input, {
            target: { value: firstItem },
        })
        const todoSubmitButton = getByRole('button', { name: /submit/i })
        fireEvent.click(todoSubmitButton)
        let todoItems = getAllByTestId(TEST_ID.todoItem)
        expect(todoItems).toHaveLength(1)
        expect(todoItems[0]).toHaveTextContent(firstItem)
        const secondItem = 'My second todo'
        fireEvent.change(input, {
            target: { value: secondItem },
        })
        fireEvent.click(todoSubmitButton)
        todoItems = getAllByTestId(TEST_ID.todoItem)
        expect(todoItems).toHaveLength(2)
        expect(todoItems[0]).toHaveTextContent(firstItem)
        expect(todoItems[1]).toHaveTextContent(secondItem)
        fireEvent.change(input, {
            target: { value: firstItem },
        })
        fireEvent.click(todoSubmitButton)
        todoItems = getAllByTestId(TEST_ID.todoItem)
        expect(todoItems).toHaveLength(3)
        expect(todoItems[0]).toHaveTextContent(firstItem)
        expect(todoItems[1]).toHaveTextContent(secondItem)
        expect(todoItems[2]).toHaveTextContent(firstItem)
        expect(asFragment()).toMatchSnapshot()
    })

    it('allows user to complete and un-complete todos', () => {
        const testTodoItems = [
            {
                id: '1',
                title: 'My Todo 1',
                createdAt: '2022-11-27T14:54:38.113Z',
                isCompleted: false,
            },
            {
                id: '2',
                title: 'My Todo 2',
                createdAt: '2022-11-28T14:54:38.113Z',
                isCompleted: false,
            },
        ]
        const { getAllByTestId } = render(
            <TodosView initialTodoItems={testTodoItems} />
        )
        const todos = getAllByTestId(TEST_ID.todoItem)
        const firstTodoCompleteButton = within(todos[0]).getByRole('button', {
            name: /complete/i,
        })
        const secondTodoCompleteButton = within(todos[1]).getByRole('button', {
            name: /complete/i,
        })
        expect(firstTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.complete
        )
        expect(secondTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.complete
        )
        fireEvent.click(firstTodoCompleteButton)
        expect(firstTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.unComplete
        )
        expect(secondTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.complete
        )
        fireEvent.click(firstTodoCompleteButton)
        expect(firstTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.complete
        )
        expect(secondTodoCompleteButton).toHaveTextContent(
            TODO_COMPLETE_BUTTON_TEXT.complete
        )
    })
})
