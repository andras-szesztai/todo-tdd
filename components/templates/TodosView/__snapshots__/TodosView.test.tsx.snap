// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosView displays submitted todos in todos list, todos with same name can be submitted 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 -translate-y-12 text-primary-dark font-medium"
    data-testid="todos-view-container"
  >
    <div
      class="flex gap-2 items-end"
    >
      <div
        class="flex flex-col gap-2"
      >
        <div
          class="ml-1"
        >
          <label
            class="font-semibold text-xl text-primary-dark cursor-pointer select-none"
            data-testid="todo-input-label"
            for="todo-input"
          >
            Todo Title
          </label>
        </div>
        <input
          class="w-60 rounded-lg border font-medium border-primary-dark focus:outline-none focus:border-accent text-base p-2 placeholder-primary-dark-disabled text-primary-dark"
          data-testid="todo-input"
          id="todo-input"
          placeholder="Enter a Todo title..."
          value=""
        />
      </div>
      <button
        class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-base p-2 rounded-lg"
        data-testid="todo-input-button"
        disabled=""
        type="button"
      >
        Submit
      </button>
    </div>
    <ul
      class="h-96 w-80 border border-primary-dark rounded-xl p-2 pr-4 overflow-y-auto bg-primary-dark-disabled flex flex-col gap-2"
      data-testid="todo-list-container"
    />
  </div>
</DocumentFragment>
`;

exports[`TodosView displays submitted todos in todos list, todos with same name can be submitted 2`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 -translate-y-12 text-primary-dark font-medium"
    data-testid="todos-view-container"
  >
    <div
      class="flex gap-2 items-end"
    >
      <div
        class="flex flex-col gap-2"
      >
        <div
          class="ml-1"
        >
          <label
            class="font-semibold text-xl text-primary-dark cursor-pointer select-none"
            data-testid="todo-input-label"
            for="todo-input"
          >
            Todo Title
          </label>
        </div>
        <input
          class="w-60 rounded-lg border font-medium border-primary-dark focus:outline-none focus:border-accent text-base p-2 placeholder-primary-dark-disabled text-primary-dark"
          data-testid="todo-input"
          id="todo-input"
          placeholder="Enter a Todo title..."
          value=""
        />
      </div>
      <button
        class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-base p-2 rounded-lg"
        data-testid="todo-input-button"
        disabled=""
        type="button"
      >
        Submit
      </button>
    </div>
    <ul
      class="h-96 w-80 border border-primary-dark rounded-xl p-2 pr-4 overflow-y-auto bg-primary-dark-disabled flex flex-col gap-2"
      data-testid="todo-list-container"
    >
      <li
        class="relative border border-primary-dark rounded-lg p-2 truncate group bg-white border-primary-dark text-primary-dark"
        data-testid="todo-item"
      >
        My first todo
        <span
          class="absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100"
        >
          <button
            class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
            data-testid="todo-complete-button"
            type="button"
          >
            Complete
          </button>
        </span>
      </li>
      <li
        class="relative border border-primary-dark rounded-lg p-2 truncate group bg-white border-primary-dark text-primary-dark"
        data-testid="todo-item"
      >
        My second todo
        <span
          class="absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100"
        >
          <button
            class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
            data-testid="todo-complete-button"
            type="button"
          >
            Complete
          </button>
        </span>
      </li>
      <li
        class="relative border border-primary-dark rounded-lg p-2 truncate group bg-white border-primary-dark text-primary-dark"
        data-testid="todo-item"
      >
        My first todo
        <span
          class="absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100"
        >
          <button
            class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
            data-testid="todo-complete-button"
            type="button"
          >
            Complete
          </button>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
